File: app/models/UserBasket.scala
package models

import reactivemongo.api.bson.{BSONDocumentReader, BSONDocumentWriter, Macros}

case class UserBasket(
                       private val basketProducts: List[BasketProduct],
                       private val totalPrice: BigDecimal
                     ) {
  def getBasketProducts = basketProducts

  def getTotalPrice = totalPrice
}

object UserBasket {
  implicit def userBasketWriter: BSONDocumentWriter[UserBasket] = Macros.writer[UserBasket]

  implicit def userBasketReader: BSONDocumentReader[UserBasket] = Macros.reader[UserBasket]
}

File: ui/src/app/_model/UserBasket.ts
import { BasketProduct } from "./BasketProduct";
import { Product } from "./Product";

export class UserBasket {
    basketProducts: BasketProduct[] = [];
    totalPrice: number = 0.0;

    constructor() {
        //
    }

    getBasketProducts() {
        return this.basketProducts;
    }

    getTotalPrice() {
        return this.totalPrice;
    }

    addProductToBasket(product: Product) {
        let idx = this.basketProducts.findIndex(bProd => bProd.product === product);
        if(idx == -1) {
            this.basketProducts.push({product: product, quantity: 1, totalPrice: product.price});
        } else {
            this.basketProducts[idx].quantity++;
            this.basketProducts[idx].totalPrice += product.price;
        }
        this.totalPrice += product.price;
    }
}